src/main/java/cbot/io/Parser.java

```diff
@@ -0,0 +1,362 @@
+                // no desc
+                throw new BadInputException("Missing event description");
+            } else if (fromIndex + fromLength >= toIndex) {
+                // no start
+                throw new BadInputException("Missing start date");
+            } else if (toIndex + toLength == this.text.length()) {
+                // no end
+                throw new BadInputException("Missing end date");
+            }
+
+            String eDesc = this.text.substring(0, fromIndex);
+            String eStartStr = this.text.substring(fromIndex + fromLength, toIndex);
+            String eEndStr = this.text.substring(toIndex + toLength);
+
+            LocalDateTime eStart = TimeStuff.parseDT(eStartStr);
+            LocalDateTime eEnd = TimeStuff.parseDT(eEndStr);
+
+            if (eStart.isAfter(eEnd)) {
+                throw new BadInputException("Hey! You have to start *before* you end...");
+            }
+
+            output = UI.say(tl.addTask(new Event(eDesc, eStart, eEnd)));
+
+            break;
+
+        case SORT:
+            if (tl.getCount() == 0) {
+                throw new PoorInputException("You have no tasks to sort :P");
+            }
+
+            tl.sort();
+            output = UI.say("Okay! I've sorted your tasks by date:\n")
+                    + UI.printMany(tl.listTasks());
+            break;
+
+        case BEFORE:
+            if (tl.getCount() == 0) {
+                throw new PoorInputException("Eh? You have no tasks to filter");
+            }
+
+            LocalDateTime bef = TimeStuff.parseDT(this.text);
+
+            ArrayList<String> arrBef = tl.listFilter(t ->
+                    t.hasTime() && t.compareTo(new Deadline("", bef)) < 0);
+
+            if (arrBef.isEmpty()) {
+                output = UI.say("You don't have any tasks before " + this.text);
+            } else {
+                output = UI.say("Here are your tasks before " + this.text + ":\n")
+                        + UI.printMany(arrBef);
+            }
+            break;
+
+        case AFTER:
+            if (tl.getCount() == 0) {
+                throw new PoorInputException("Eh? You have no tasks to filter");
+            }
+
+            LocalDateTime aft = TimeStuff.parseDT(this.text);
+
+            ArrayList<String> arrAft = tl.listFilter(t ->
+                    t.hasTime() && t.compareTo(new Deadline("", aft)) > 0);
+
+            if (arrAft.isEmpty()) {
+                output = UI.say("You don't have any tasks after " + this.text);
+            } else {
+                output = UI.say("Here are your tasks after " + this.text + ":\n")
+                        + UI.printMany(arrAft);
+            }
+            break;
+
+        case FILTER:
+            if (tl.getCount() == 0) {
+                throw new PoorInputException("Eh? You have no tasks to filter");
+            }
+
+            String msg;
+            ArrayList<String> arrFilter;
+
+            switch (this.text.toLowerCase()) {
```

Comment:
Is there a point here in including cases with no behaviour if they are gonna fallthrough anyways?

Reply:
Oh! These are just to give the user a bit more flexibility in how they call the filter. For example, both `filter deadline` and `filter dl` (which do the same thing) are accepted and understood by Cbot.
If you know a neater way of doing this, please do let me know!